pwndbg/heap/ptmalloc.py

```diff
@@ -547,11 +487,906 @@ def largebin_index(self, sz):
+        """
+        region = self.get_region(addr) if addr else self.get_region(self.get_arena()['top'])
+
+        # Occasionally, the [heap] vm region and the actual start of the heap are
+        # different, e.g. [heap] starts at 0x61f000 but mp_.sbrk_base is 0x620000.
+        # Return an adjusted Page object if this is the case.
+        page = pwndbg.memory.Page(0, 0, 0, 0)
+        sbrk_base = int(self.mp['sbrk_base'])
+        if region == self.get_region(sbrk_base):
+            if sbrk_base != region.vaddr:
+                page.vaddr = sbrk_base
+                page.memsz = region.memsz - (sbrk_base - region.vaddr)
+                return page
+        return region
+
     def is_initialized(self):
         addr = pwndbg.symbol.address('__libc_malloc_initialized')
         if addr is None:
             addr = pwndbg.symbol.address('__malloc_initialized')
         return pwndbg.memory.s32(addr) > 0
 
-    def libc_has_debug_syms(self):
-        return pwndbg.symbol.address('global_max_fast') is not None
+
+def inform_report_issue(symbols):
+    print(
+        message.notice("Symbol: ") +
+        message.hint(symbols) +
+        message.notice(" could not be found via heuristics for ") +
+        message.hint(pwndbg.arch.current) +
+        message.notice(" with GLIBC version ") +
+        message.hint(".".join(str(i) for i in pwndbg.glibc.get_version()))
+    )
+    print(
+        message.notice("If that is an issue, you can report it on https://github.com/pwndbg/pwndbg/issues\n"
+                       "If possible, please provide your libc.so.6 when reporting the issue. Thanks!"
+                       "(Please don't forget to search if it hasn't been reported before)\n"
+                       "To generate the report and open a browser, you may run ") +
+        message.hint("`bugreport --run-browser`") +
+        message.notice("\nPS: Pull requests are welcome")
+    )
+
+
+def initialize_structs(method):
+    @wraps(method)
+    def wrapper(self, *args, **kwargs):
+        # If structs module for HeuristicHeap is None, try to initialize it.
+        if self._structs_module is None:
+            self._structs_module = importlib.reload(importlib.import_module('pwndbg.heap.structs'))
+        return method(self, *args, **kwargs)
+    return wrapper
+
+
+class HeuristicHeap(Heap):
+    def __init__(self):
+        super().__init__()
+        self._thread_arena_offset = None
+        self._thread_cache_offset = None
+        self._structs_module = None
+        self._errno_offset = None
+
+    @property
+    def main_arena(self):
+        main_arena_via_config = int(str(pwndbg.config.main_arena), 0)
+        if main_arena_via_config > 0:
+            return self.malloc_state(main_arena_via_config)
+        # TODO/FIXME: These are quite dirty, we should find a better way to do this
+        if not self._main_arena_addr:
+            if pwndbg.glibc.get_version() < (2, 34) and pwndbg.arch.current != "arm":
```

Comment:
Just to be sure: what happens for "arm"? Is arm 32-bit fine?

Reply:
Hmm, GEF's heuristic for `main_arena` seems to work fine for i386, x86-64, aarch64 with GLIBC < 2.34, but I'm not sure what happened on the heuristic of GEF for 32-bit arm actually, so I decided to use our version of heuristic.
And just guessing, maybe it's because GEF got the wrong `malloc_state` size for arm 32-bit?
You might reproduce it with a normal GLIBC 2.31 which is installed via `apt-get install arm-linux-gnueabihf-gcc`.

Here is a example, if we try the code from https://github.com/hugsy/gef/blob/dev/gef.py#L1180

```
gef➤  pi
>>> ctypes.sizeof(GlibcArena.malloc_state_t())
1116
>>> malloc_hook_addr = parse_address("(void *)&__malloc_hook")
>>> struct_size = ctypes.sizeof(GlibcArena.malloc_state_t())
>>> hex(malloc_hook_addr - gef.arch.ptrsize - struct_size)
'0x3f7bf76c'
>>>
gef➤  ptype/o main_arena
/* offset      |    size */  type = struct malloc_state {
/*      0      |       4 */    __libc_lock_t mutex;
/*      4      |       4 */    int flags;
/*      8      |       4 */    int have_fastchunks;
/*     12      |      40 */    mfastbinptr fastbinsY[10];
/*     52      |       4 */    mchunkptr top;
/*     56      |       4 */    mchunkptr last_remainder;
/*     60      |    1016 */    mchunkptr bins[254];
/*   1076      |      16 */    unsigned int binmap[4];
/*   1092      |       4 */    struct malloc_state *next;
/*   1096      |       4 */    struct malloc_state *next_free;
/*   1100      |       4 */    size_t attached_threads;
/*   1104      |       4 */    size_t system_mem;
/*   1108      |       4 */    size_t max_system_mem;

                               /* total size (bytes): 1112 */
                             }
gef➤  p &main_arena
$1 = (struct malloc_state *) 0x3f7bf76c <main_arena>
```
As you can see, idk why, GEF's heuristic got a correct `main_arena` address with a wrong size of `malloc_state` lol
But because our size is correct, it will failed.